src/composables/useAsync.ts

```diff
@@ -0,0 +1,91 @@
+/**
+ * @license Copyright (c) 2003-2024, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+import { computed, ref, watchEffect, type ComputedRef, type Ref } from 'vue';
+import { uid } from '@ckeditor/ckeditor5-integrations-common';
+
+/**
+ * A composable that executes an async function and provides the result.
+ *
+ * @param asyncFunc The async function to execute.
+ * @returns The result of the async function.
+ * @example
+ *
+ * ```ts
+ * const { loading, data, error } = useAsync( async () => {
+ * 	const response = await fetch( 'https://api.example.com/data' );
+ * 	return response.json();
+ * } );
+ * ```
+ */
+export const useAsync = <R>(
+	asyncFunc: () => Promise<R>
+): AsyncComposableResult<R> => {
+	// The UUID of the last query to prevent race conditions between multiple queries.
+	const lastQueryUUID = ref<string | null>( null );
+
+	// The error thrown by the async function.
+	const error = ref<Error | null>( null );
+
+	// The data returned by the async function.
+	const data: Ref<R | null> = ref( null );
+
+	// Whether the async function is currently loading.
+	const loading = computed( () => lastQueryUUID.value !== null );
+
+	// Execute the async function and update the result.
+	watchEffect( async () => {
+		const currentQueryUID = uid();
+
+		lastQueryUUID.value = currentQueryUID;
+		data.value = null;
+		error.value = null;
+
+		const shouldDiscardQuery = () => lastQueryUUID.value !== currentQueryUID;
+
+		try {
+			const result = await asyncFunc();
+
+			if ( !shouldDiscardQuery() ) {
+				data.value = result;
+			}
+		} catch ( err: any ) {
+			if ( !shouldDiscardQuery() ) {
+				error.value = err;
+			}
+		} finally {
+			if ( !shouldDiscardQuery() ) {
+				lastQueryUUID.value = null;
+			}
+		}
+	} );
+
+	return {
+		loading,
+		data,
+		error
```

Comment:
```suggestion
		loading: shallowReadonly( loading ),
		data: shallowReadonly( data ),
		error: shallowReadonly (error )
```

This is to prevent reassignment of these refs from outside and triggering the `watchEffect`.

Please also add tests checking this.

Reply:
Good idea, thanks!